### PR TITLE
feat(backups): add support for local, dropbox and gdrive strategies for backups

### LINKED ISSUE #1971: I should be able to enable automated backups of channel state data
## Detailed Description

I should be able to enable automated backups of channel state data to Dropbox so that I can more easily recover channel state data in the event of needing to restore.

## Context

Static Channel Backups (new lnd feature)

## Possible Implementation

Build out Dropbox integration and provide an easy way for users to enable automated, realtime backups of channel state data to a designated Dropbox account.

### LINKED ISSUE #2214: Implement local backup strategy option for backup setup
## Detailed Description
User should be able to select local SCB  strategy when setting up backup during onboarding.

### LINKED ISSUE #2215: Implement dropbox backup strategy option for backup setup
## Detailed Description

User should be able to select dropbox SCB strategy when setting up backup during onboarding.

### PR DESCRIPTION
**Pull request by korhaliv**
## Description:
resolves #1971
resolves #2214
resolves #2215

## How Has This Been Tested?
Manually

## Screenshots (if appropriate):
image

## Types of changes:
New feature

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have reviewed and updated the documentation accordingly.
- [x] I have read the _CONTRIBUTING_ document.
- [ ] I have added tests to cover my changes where needed.
- [ ] All new and existing tests passed.
- [x] My commits have been squashed into a concise set of changes.

### CODE REVIEW

File: electron/walletBackup/dropbox/dbApi.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    },
+  })
+
+  const authUrl = createAuthUrl(clientId, redirectUrl)
+
+  return new Promise((resolve, reject) => {
+    const handleNavigation = url => {
+      const { hash, query } = parse(url, true)
+      if (query && query.error) {
+        reject(new Error(`There was an error: ${query.error}`))
```
[the remark this answers is not in the document]
Comment (korhaliv) [reply electron/walletBackup/dropbox/dbApi.js]: nice

File: electron/walletBackup/dropbox/index.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import config from 'config'
+
```
[the remark this answers is not in the document]
Comment (korhaliv) [reply electron/walletBackup/dropbox/index.js]: done

File: electron/walletBackup/dropbox/index.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import chainify from '@zap/utils/chainify'
+import { mainLog } from '@zap/utils/log'
+import TokenBasedBackupService from '../base/TokenBasedBackupService'
+import createClient from './db'
+
+class BackupService extends TokenBasedBackupService {
+  /**
+   * Setups gdrive service for usage. This method must be called before calling any other methods
```
[the remark this answers is not in the document]
Comment (korhaliv) [reply electron/walletBackup/dropbox/index.js]: done

File: electron/walletBackup/local/index.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+   *
+   * @readonly
+   * @memberof BackupService
+   */
+  get name() {
+    return 'local'
+  }
+
+  /**
+   * This service doesn't use tokens and  doesn't emit token events
```
[the remark this answers is not in the document]
Comment (korhaliv) [reply electron/walletBackup/local/index.js]: done

File: electron/walletBackup/local/index.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  get isUsingTokens() {
+    return false
+  }
+}
+
+// singleton backup service
+
+let backupService
+
+export default function getBackupService() {
```
[the remark this answers is not in the document]
Comment (korhaliv) [reply electron/walletBackup/local/index.js]: nice suggestion

File: electron/walletBackup/service.js [hunk 6]
```diff
@@ -6,28 +6,40 @@ export default function createBackupService(mainWindow) {
   // helper func to send messages to the renderer process
   const send = (msg, params) => mainWindow.webContents.send(msg, params)
 
+  // setups ipc to listen for backup service token updates and forward updates to
```
[the remark this answers is not in the document]
Comment (korhaliv) [reply electron/walletBackup/service.js]: done

File: renderer/components/UI/OpenDialogInput.js [hunk 7]
```diff
@@ -39,6 +40,7 @@ const OpenDialogInput = props => {
 
 OpenDialogInput.propTypes = {
   fieldApi: PropTypes.object.isRequired,
+  mode: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (korhaliv) [reply renderer/components/UI/OpenDialogInput.js]: renamed

File: electron/walletBackup/local/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrfelton) [reply electron/walletBackup/local/index.js]: still a double space here

[the remark this answers is not in the document]
Comment (korhaliv) [reply electron/walletBackup/local/index.js]: there was one more place with double space so I've changed there but not here :)
updated

File: renderer/store/db.js [hunk 8]
```diff
@@ -138,6 +138,9 @@ export const getDb = name => {
     // Save the migrated config overrides.
     await tx.settings.put({ key: 'config', value: newConfig })
   })
+  db.version(5).stores({
```
Comment (mrfelton) [inline renderer/store/db.js]: No need for new db table, right @korhaliv ?

Comment (korhaliv) [reply renderer/store/db.js]: Right, bad merge result

File: package.json [hunk 9]
```diff
@@ -349,9 +351,11 @@
     "lodash.partition": "4.6.0",
     "lodash.pick": "4.4.0",
     "lodash.set": "4.3.2",
+    "lodash.snakecase": "4.1.1",
```
Comment (mrfelton) [inline package.json]: Looks like this got added back in too

File: renderer/components/Onboarding/Steps/ConnectionType.js [hunk 10]
```diff
@@ -2,22 +2,15 @@ import React from 'react'
...
+import Container from './components/Container'
 import messages from './messages'
 
 const ConnectionTypeItem = withFieldApi('connectionType')(BaseConnectionTypeItem)
 
-const Container = styled(Flex)`
-  position: absolute;
-  top: -30px;
-  bottom: 0;
-  visibility: ${props => (props.lndConnect ? 'hidden' : 'visible')};
```
Comment (mrfelton) [inline renderer/components/Onboarding/Steps/ConnectionType.js]: Removing the `visibility` based on `props.lndConnect` causes the steps to flash briefly when using an lnd connect link.

Comment (korhaliv) [reply renderer/components/Onboarding/Steps/ConnectionType.js]: resolved

File: package.json (hunk 9 above)
Comment (korhaliv) [reply package.json]: removed

**Comment by korhaliv** [comment]
@mrfelton csp for prod updated

**Pull request merged**